# Actividad 19. Enjambre de partículas

In [25]:
from random import random

# funcion objetivo a minimizar
def funcion(x, y):
    sum1 = x**2 * (4 - 2.1*x**2 + (x**4)/3.0)
    sum2 = x*y
    sum3 = y**2 * (-4 + 4*y**2)
    return sum1 + sum2 + sum3

# genera numero aleatorio en rango
def aleatorio(inf, sup):
    return random() * (sup - inf) + inf

# crea una particula como diccionario
def crear_particula():
    p = {}
    p["x"] = aleatorio(infx, supx)
    p["y"] = aleatorio(infy, supy)
    p["vx"] = aleatorio(infx/ajusteV, supx/ajusteV)
    p["vy"] = aleatorio(infy/ajusteV, supy/ajusteV)
    p["xLoc"] = p["x"]
    p["yLoc"] = p["y"]
    p["valorLoc"] = funcion(p["x"], p["y"])
    return p

# actualiza la velocidad
def actualizar_velocidad(p, xGlob, yGlob):
    cogX = cognitiva * random() * (p["xLoc"] - p["x"])
    socX = social * random() * (xGlob - p["x"])
    p["vx"] = inercia * p["vx"] + cogX + socX
    cogY = cognitiva * random() * (p["yLoc"] - p["y"])
    socY = social * random() * (yGlob - p["y"])
    p["vy"] = inercia * p["vy"] + cogY + socY

# actualiza la posicion
def actualizar_posicion(p):
    p["x"] = p["x"] + p["vx"]
    p["y"] = p["y"] + p["vy"]

# limita la posicion al espacio de soluciones
def limitar_posicion(p):
    p["x"] = max(p["x"], infx)
    p["x"] = min(p["x"], supx)
    p["y"] = max(p["y"], infy)
    p["y"] = min(p["y"], supy)

# actualiza mejor valor local
def actualizar_mejor_local(p):
    valor = funcion(p["x"], p["y"])
    if valor < p["valorLoc"]:
        p["xLoc"] = p["x"]
        p["yLoc"] = p["y"]
        p["valorLoc"] = valor

# algoritmo principal PSO
def enjambre_particulas(particulas, iteraciones, reduccionInercia):
    global inercia

    mejor = min(particulas, key=lambda p: p["valorLoc"])
    xGlob = mejor["xLoc"]
    yGlob = mejor["yLoc"]
    valorGlob = mejor["valorLoc"]

    for _ in range(iteraciones):
        for p in particulas:
            actualizar_velocidad(p, xGlob, yGlob)
            actualizar_posicion(p)
            limitar_posicion(p)
            actualizar_mejor_local(p)

        mejor = min(particulas, key=lambda p: p["valorLoc"])
        if mejor["valorLoc"] < valorGlob:
            xGlob = mejor["xLoc"]
            yGlob = mejor["yLoc"]
            valorGlob = mejor["valorLoc"]

        inercia *= reduccionInercia

    return xGlob, yGlob, valorGlob

# parametros del modelo
inercia = 1.4
cognitiva = 2.0
social = 2.0

# limites del espacio
infx = -10.0
supx = 10.0
infy = -10.0
supy = 10.0

# ajuste velocidad inicial
ajusteV = 100.0

# parametros de ejecucion
nParticulas = 10
iteraciones = 100
redInercia = 0.9

# crea enjambre inicial
particulas = [crear_particula() for _ in range(nParticulas)]

# ejecuta algoritmo
print(enjambre_particulas(particulas, iteraciones, redInercia))

(0.08984201185608622, -0.7126564006468207, -1.0316284534898774)


In [27]:
from random import random

# Función que se quiere minimizar
def funcion(x, y):
    sum1 = x**2 * (4 - 2.1*x**2 + x**4/3.0)
    sum2 = x*y
    sum3 = y**2 * (-4 + 4*y**2)
    return sum1 + sum2 + sum3

# Devuelve un número aleatorio dentro de un rango con
# distribución uniforme (proporcionada por random)
def aleatorio(inf, sup):
    return random()*(sup-inf) + inf

# Clase que representa una partícula individual y que facilita
# las operaciones necesarias
class Particula:
    # Algunos atributos de clase (comunes a todas las partículas)
    # Parámetros para actualizar la velocidad
    inercia = 1.4
    cognitiva = 2.0
    social = 2.0
    # Límites del espacio de soluciones
    infx = -10.0
    supx = 10.0
    infy = -10.0
    supy = 10.0

    # Factor de ajuste de la velocidad inicial
    ajusteV = 100.0

    # Crea una partícula dentro de los límites indicados
    def __init__(self):
        self.x = aleatorio(Particula.infx, Particula.supx)
        self.y = aleatorio(Particula.infy, Particula.supy)
        self.vx = aleatorio(Particula.infx/Particula.ajusteV,
                            Particula.supx/Particula.ajusteV)
        self.vy = aleatorio(Particula.infy/Particula.ajusteV,
                            Particula.supy/Particula.ajusteV)
        self.xLoc = self.x
        self.yLoc = self.y
        self.valorLoc = funcion(self.x, self.y)

    # Actualiza la velocidad de la partícula
    def actualizaVelocidad(self, xGlob, yGlob):
        cogX = Particula.cognitiva*random()*(self.xLoc-self.x)
        socX = Particula.social*random()*(xGlob-self.x)
        self.vx = Particula.inercia*self.vx + cogX + socX
        cogY = Particula.cognitiva*random()*(self.yLoc-self.y)
        socY = Particula.social*random()*(yGlob-self.y)
        self.vy = Particula.inercia*self.vy + cogY + socY

    # Actualiza la posición de la partícula
    def actualizaPosicion(self):
        self.x = self.x + self.vx
        self.y = self.y + self.vy

        # Debe mantenerse dentro del espacio de soluciones
        self.x = max(self.x, Particula.infx)
        self.x = min(self.x, Particula.supx)
        self.y = max(self.y, Particula.infy)
        self.y = min(self.y, Particula.supy)

        # Si es inferior a la mejor, la adopta como mejor
        valor = funcion(self.x, self.y)
        if valor < self.valorLoc:
            self.xLoc = self.x
            self.yLoc = self.y
            self.valorLoc = valor

# Mueve un enjambre de partículas durante las iteraciones indicadas.
# Devuelve las coordenadas y el valor del mínimo obtenido.
def enjambreParticulas(particulas, iteraciones, reduccionInercia):

    # Registra la mejor posición global y su valor
    mejorParticula = min(particulas, key=lambda p:p.valorLoc)
    xGlob = mejorParticula.xLoc
    yGlob = mejorParticula.yLoc
    valorGlob = mejorParticula.valorLoc

    # Bucle principal de simulación
    for iter in range(iteraciones):
        # Actualiza la velocidad y posición de cada partícula
        for p in particulas:
            p.actualizaVelocidad(xGlob, yGlob)
            p.actualizaPosicion()

        # Hasta que no se han movido todas las partículas no se
        # actualiza el mínimo global, para simular que todas se
        # mueven a la vez
        mejorParticula = min(particulas, key=lambda p:p.valorLoc)
        if mejorParticula.valorLoc < valorGlob:
            xGlob = mejorParticula.xLoc
            yGlob = mejorParticula.yLoc
            valorGlob = mejorParticula.valorLoc

        # Finalmente se reduce la inercia de las partículas
        Particula.inercia *= reduccionInercia

    return (xGlob, yGlob, valorGlob)

# Parámetros del problema
nParticulas = 10
iteraciones = 100
redInercia = 0.9

# Genera un conjunto inicial de partículas
particulas = [Particula() for i in range(nParticulas)]

# Ejecuta el algoritmo del enjambre de partículas
print(enjambreParticulas(particulas, iteraciones, redInercia))


(0.08984201217183363, -0.7126564052807144, -1.0316284534898774)


### Comparación de resultados

En la implementación estructurada se obtuvo el resultado (0.08984201185608622, -0.7126564006468207, -1.0316284534898774), mientras que en la implementación del libro se obtuvo (0.08984201217183363, -0.7126564052807144, -1.0316284534898774).

Al compararlos, se puede ver que prácticamente son iguales. Tanto las coordenadas como el valor mínimo coinciden casi exactamente, y las pequeñas diferencias que aparecen son muy mínimas, probablemente por temas de redondeo o por la naturaleza aleatoria del algoritmo.

Esto indica que ambas implementaciones están funcionando correctamente y llegan al mismo resultado. En general, se puede concluir que la versión estructurada replica bien el comportamiento de la implementación original del libro.